ndk/src/event.rs

```diff
@@ -47,12 +47,20 @@ pub enum Source {
     Any = ffi::AINPUT_SOURCE_ANY,
 }
 
+impl Source {
+    pub fn class(self) -> SourceClass {
+        let class = u32::from(self) & ffi::AINPUT_SOURCE_CLASS_MASK;
+        // TODO: Don't use None
+        class.try_into().unwrap_or(SourceClass::None)
+    }
+}
+
 /// An enum representing the class of an [`InputEvent`] source.
 ///
 /// See [the NDK docs](https://developer.android.com/ndk/reference/group/input#anonymous-enum-35)
 #[derive(Debug, Clone, Copy, PartialEq, Eq, TryFromPrimitive, IntoPrimitive)]
 #[repr(u32)]
-enum Class {
+pub enum SourceClass {
```

Comment:
It's slightly awkward / inconsistent that `_SOURCE_CLASS_NAVIGATION` is named `_SOURCE_CLASS_TRACKBALL` in the SDK, and that's the name used in `android-activity`. Probably makes sense to go with the NDK naming here though.

In general though, the SourceClass is a bitmask and I'm not sure that the class is strictly mutually exclusive. In `android-activity` this is bound as a `SourceFlags` bitmask but only exposed via `Source` via methods like `Source::is_button_class()` - I'd maybe be tempted to do something similar here too.

Reply:
Yeah, "navigation" doesn't mean much. Do you want me to solve this with an extra doc alias?

Reply:
You are totally light, this must be `bitflags`!

C definition:

```c
/**
 * Input source masks.
 *
 * Refer to the documentation on android.view.InputDevice for more details about input sources
 * and their correct interpretation.
 */
enum {
    /** mask */
    AINPUT_SOURCE_CLASS_MASK = 0x000000ff,

    /** none */
    AINPUT_SOURCE_CLASS_NONE = 0x00000000,
    /** button */
    AINPUT_SOURCE_CLASS_BUTTON = 0x00000001,
    /** pointer */
    AINPUT_SOURCE_CLASS_POINTER = 0x00000002,
    /** navigation */
    AINPUT_SOURCE_CLASS_NAVIGATION = 0x00000004,
    /** position */
    AINPUT_SOURCE_CLASS_POSITION = 0x00000008,
    /** joystick */
    AINPUT_SOURCE_CLASS_JOYSTICK = 0x00000010,
};
```